pkg/virt-controller/services/rendervolumes.go

```diff
@@ -0,0 +1,572 @@
+				continue
+			}
+			volumeName := secretName + "-access-cred"
+			renderer.podVolumes = append(renderer.podVolumes, k8sv1.Volume{
+				Name: volumeName,
+				VolumeSource: k8sv1.VolumeSource{
+					Secret: &k8sv1.SecretVolumeSource{
+						SecretName: secretName,
+					},
+				},
+			})
+			renderer.podVolumeMounts = append(renderer.podVolumeMounts, k8sv1.VolumeMount{
+				Name:      volumeName,
+				MountPath: filepath.Join(config.SecretSourceDir, volumeName),
+				ReadOnly:  true,
+			})
+		}
+		return nil
+	}
+}
+
+func withSidecarVolumes(hookSidecars hooks.HookSidecarList) VolumeRendererOption {
+	return func(renderer *VolumeRenderer) error {
+		if len(hookSidecars) != 0 {
+			renderer.podVolumes = append(renderer.podVolumes, k8sv1.Volume{
+				Name: hookSidecarSocks,
+				VolumeSource: k8sv1.VolumeSource{
+					EmptyDir: &k8sv1.EmptyDirVolumeSource{},
+				},
+			})
+			renderer.podVolumeMounts = append(renderer.podVolumeMounts, k8sv1.VolumeMount{
+				Name:      hookSidecarSocks,
+				MountPath: hooks.HookSocketsSharedDirectory,
+			})
+		}
+		return nil
+	}
+}
+
+func withHugepages() VolumeRendererOption {
+	return func(renderer *VolumeRenderer) error {
+		renderer.podVolumes = append(renderer.podVolumes, k8sv1.Volume{
+			Name: "hugepages",
+			VolumeSource: k8sv1.VolumeSource{
+				EmptyDir: &k8sv1.EmptyDirVolumeSource{
+					Medium: k8sv1.StorageMediumHugePages,
+				},
+			},
+		})
+		renderer.podVolumeMounts = append(renderer.podVolumeMounts, k8sv1.VolumeMount{
+			Name:      "hugepages",
+			MountPath: filepath.Join("/dev/hugepages"),
+		})
+		return nil
+	}
+}
+
+func withHotplugSupport(hotplugDiskDir string) VolumeRendererOption {
+	return func(renderer *VolumeRenderer) error {
+		prop := k8sv1.MountPropagationHostToContainer
+		renderer.podVolumeMounts = append(renderer.podVolumeMounts, k8sv1.VolumeMount{
+			Name:             hotplugDisks,
+			MountPath:        hotplugDiskDir,
+			MountPropagation: &prop,
+		})
+		renderer.podVolumes = append(renderer.podVolumes, k8sv1.Volume{
+			Name: hotplugDisks,
+			VolumeSource: k8sv1.VolumeSource{
+				EmptyDir: &k8sv1.EmptyDirVolumeSource{},
+			},
+		})
+		return nil
+	}
+}
+
+func imgPullSecrets(volumes ...v1.Volume) []k8sv1.LocalObjectReference {
+	var imagePullSecrets []k8sv1.LocalObjectReference
+	for _, volume := range volumes {
+		if volume.ContainerDisk != nil && volume.ContainerDisk.ImagePullSecret != "" {
+			imagePullSecrets = appendUniqueImagePullSecret(imagePullSecrets, k8sv1.LocalObjectReference{
+				Name: volume.ContainerDisk.ImagePullSecret,
+			})
+		}
+	}
+	return imagePullSecrets
+}
+
+func serviceAccount(volumes ...v1.Volume) string {
+	for _, volume := range volumes {
+		if volume.ServiceAccount != nil {
+			return volume.ServiceAccount.ServiceAccountName
+		}
+	}
+	return ""
+}
+
+func addPVCToLaunchManifest(pvcStore cache.Store, volume v1.Volume, claimName string, namespace string, volumeMounts *[]k8sv1.VolumeMount, volumeDevices *[]k8sv1.VolumeDevice) error {
```

Comment:
what do you think about making this function 
```
func (vr *VolumeRenderer) addPVCToLaunchManifest(pvcStore cache.Store, volume v1.Volume, claimName string)
``` 
It "knows" how to add pvc to launch manifest and it needs three members of VolumeRenderer struct, so I think it belongs to VolumeRenderer. 

Maybe there are similar functions like that?

Reply:
Makes sense, good suggestion.

Reply:
Done.